flink-streaming-java/src/test/java/org/apache/flink/streaming/util/AbstractStreamOperatorTestHarnessTest.java

```diff
@@ -52,4 +55,24 @@ public void testInitializeAfterOpenning() throws Throwable {
 		result.open();
 		result.initializeState(new OperatorSubtaskState());
 	}
+
+	@Test
+	public void testSetTtlTimeProvider() throws Exception {
+		AbstractStreamOperatorTestHarness<Integer> result;
+		AbstractStreamOperator operator = new AbstractStreamOperator<Integer>() {};
+		result =
+			new AbstractStreamOperatorTestHarness<>(
+				operator,
+				1,
+				1,
+				0);
+		result.config.setStateKeySerializer(IntSerializer.INSTANCE);
+
+		long expectedTimeStamp = 42;
+		result.setTtlTimeProvider(() -> expectedTimeStamp);
+		result.initializeState(new OperatorSubtaskState());
+		result.open();
+		Assert.assertEquals(expectedTimeStamp,
+			((AbstractKeyedStateBackend<?>) operator.getKeyedStateBackend()).getTtlTimeProvider().currentTimestamp());
```

Comment:
I would avoid exposing `getTtlTimeProvider` in `AbstractKeyedStateBackend`, although we annotate with `VisibleForTesting`, it still becomes visible for users.
The working of custom `TtlTimeProvider` can be tested e.g. by creating TTL states and checking that they expire according the custom timestamp. Also this will be tested by other tests which rely on this injection of custom `TtlTimeProvider`.

Reply:
This sounds good, and I will update this test.